### DISCUSSION TITLE
Go Target rethink

### DISCUSSION THREAD

**Discussion by parrt** [antlr/antlr4#3812] [Ideas]
Start a new conversation about the Go target here from https://github.com/antlr/antlr4/issues/3718#issuecomment-1212662270

@jimidle has agreed to take a look

**Comment by parrt** [comment]
We should definitely think about the visitor pattern to see if it's correct:

https://github.com/antlr/antlr4/issues/2504

I also remember either getting a crash or no error message upon syntax error when I tried Go target.

**Comment by parrt** [comment]
I think there were some data race fixes also for Go: https://github.com/antlr/antlr4/pull/3137  so the multi threaded correctness might be OK.  @Joakker would you care to comment on the current thread safety of Go?

**Comment by parrt** [comment]
Another thoughtful analysis about the various interfaces in Go target: https://github.com/antlr/antlr4/issues/1843

**Comment by kaby76** [comment]
I recommend reading what I said in:

* #3452 where two different grammars are placed in one directory, e.g., grammars-v4/csharp/, and name collisions are created because parsers are created with global variable names but scoped by package name. It complicates the design of a program that uses two parsers. Suggest renaming global variables so that the recognizer name is included. I could work around it, but it's a hassle. Take the code in grammars-v4/csharp and just try to write a program that works and see what I mean. See also to #3264. Everything is declared globally, scoped by package.
* #3446 where it's impossible to call "base class" (in this design, really a "base struct") initialization of a recognizer. Every field in "base classes" have to use "zero initialize". Try porting grammars-v4/sql/plsql/ to Go.
* #3443

It feels like the Go target wasn't really tested with the superClass option, trying the "target-agnostic format" model.

**Reply by KvanTTT** [reply]
> It feels like the Go target wasn't really tested with the superClass option, trying the "target-agnostic format" model.

It's the right time to create new tests for `superClass` option.

**Reply by parrt** [reply]
oh wow. We don't have any? zoiks.

**Reply by parrt** [reply]
hmm... Global variables in Go target? yuck.

**Reply by parrt** [reply]
Very useful links, thanks @kaby76   Hopefully @jimidle can address these as well.

**Comment by parrt** [comment]
Looks like we can close as @jimidle has corrected most stuff.
